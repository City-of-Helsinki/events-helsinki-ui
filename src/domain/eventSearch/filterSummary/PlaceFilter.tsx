import React from "react";
import { useTranslation } from "react-i18next";

import FilterButton, {
  FilterType
} from "../../../common/components/filterButton/FilterButton";
import { usePlaceDetailsQuery } from "../../../generated/graphql";
import useLocale from "../../../hooks/useLocale";
import getLocalisedString from "../../../util/getLocalisedString";

interface Props {
  id: string;
  onRemove: (value: string, type: FilterType) => void;
}

const PublisherFilter: React.FC<Props> = ({ id, onRemove }) => {
  const { t } = useTranslation();
  const locale = useLocale();
  const { data, loading } = usePlaceDetailsQuery({
    variables: { id }
  });

  return loading ? (
    <FilterButton
      onRemove={onRemove}
      text={t("commons.loading")}
      type="place"
      value={id}
    />
  ) : data && data.placeDetails.name ? (
    <FilterButton
      onRemove={onRemove}
      text={getLocalisedString(data.placeDetails.name, locale)}
      type="place"
      value={id}
    />
  ) : null;
};

export default PublisherFilter;
