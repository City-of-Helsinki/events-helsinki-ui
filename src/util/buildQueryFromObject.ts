/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
import forEach from 'lodash/forEach';
import isArray from 'lodash/isArray';
import isEmpty from 'lodash/isEmpty';
import isNil from 'lodash/isNil';
import isNumber from 'lodash/isNumber';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const buildQueryFromObject = (obj: { [key: string]: any }) => {
  const query: string[] = [];

  forEach(obj, (filter, key) => {
    if (!isEmpty(filter) || isNumber(filter) || typeof filter === 'boolean') {
      if (isArray(filter)) {
        const items: Array<string | number> = [];

        forEach(filter, (item: string | number) => {
          items.push(encodeURIComponent(item));
        });

        query.push(`${key}=${items.join(',')}`);
      } else if (!isNil(filter)) {
        query.push(`${key}=${encodeURIComponent(filter)}`);
      }
    }
  });

  return query.length ? `?${query.join('&')}` : '';
};

export default buildQueryFromObject;
