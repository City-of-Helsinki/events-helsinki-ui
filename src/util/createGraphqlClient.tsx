import * as Sentry from "@sentry/browser";
import { InMemoryCache } from "apollo-cache-inmemory";
import ApolloClient from "apollo-client";
import { ApolloLink } from "apollo-link";
import { BatchHttpLink } from "apollo-link-batch-http";
import { onError } from "apollo-link-error";
import { createPersistedQueryLink } from "apollo-link-persisted-queries";
import get from "lodash/get";

export default function createClient(uri: string) {
  const customFetch = (batchUri: string, options: RequestInit) => {
    // TODO: Inject authentication token to requests when Tunnistamo is added
    return fetch(uri, options);
  };

  const errorLink = onError(({ graphQLErrors, networkError }) => {
    if (graphQLErrors) {
      graphQLErrors.forEach(({ message, locations, path }) => {
        const errorMessage = `[GraphQL error]: Message: ${message}, Location: ${locations}, Path: ${path}`;
        Sentry.captureMessage(errorMessage);
      });
    }

    if (networkError) {
      Sentry.captureMessage("Network error");
    }
  });

  const link = ApolloLink.from([
    createPersistedQueryLink(),
    errorLink,
    new BatchHttpLink({
      batchMax: 1,
      fetch: customFetch
    })
  ]);

  return new ApolloClient({
    cache: new InMemoryCache({
      addTypename: true
    }).restore(get(window, "__APOLLO_STATE__")),
    link
  });
}
