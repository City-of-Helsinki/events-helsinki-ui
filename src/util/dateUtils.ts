import { format as formatDateStr } from 'date-fns';
import isAfter from 'date-fns/isAfter';
import isValid from 'date-fns/isValid';
import { enGB as en, fi } from 'date-fns/locale';
import parse from 'date-fns/parse';
import get from 'lodash/get';

import sv from './date-fns/locale/sv';

const locales = { en, fi, sv };

export const formatDate = (
  date: Date | number | null,
  format = 'd.M.yyyy',
  locale = 'fi'
): string => {
  if (!date) {
    return '';
  }

  return formatDateStr(date, format, {
    locale: get(locales, locale),
  }).trim();
};

export const isValidDateString = (date: string) => {
  return isValidDate(parse(date, 'd.M.yyyy', new Date()));
};

export const parseDate = (date: string) => {
  return parse(date, 'd.M.yyyy', new Date());
};

/**
 * Test is date valid
 */
const isValidDate = (date: Date): boolean =>
  isValid(date) && isAfter(date, new Date('1000-01-01'));

/**
 * Test is entered string a date string in Finnish format without dots (e.g. 31122019)
 */
const isShortDateStr = (str: string) =>
  str.length === 8 && /^[0-9.]+$/.test(str);

const getShortDateStr = (str: string): string =>
  [str.substring(0, 2), str.substring(2, 4), str.substring(4, 9)].join('.');

/**
 * Get date object from valid Finnish date string
 */
const getParsedDate = (value: string): Date =>
  parse(value, 'd.M.yyyy', new Date(), { locale: fi });

/**
 * Convert string in Finnish date format (e.g. 31.12.2019) or in format without dots (e.g. 31122019) to Date object
 */
export const convertFinnishDateStrToDate = (str: string): Date | null => {
  let parsedDate = getParsedDate(str);

  if (isValidDate(parsedDate)) {
    return parsedDate;
  } else if (isShortDateStr(str)) {
    const dateStr = getShortDateStr(str);

    parsedDate = getParsedDate(dateStr);

    if (isValidDate(parsedDate)) {
      return parsedDate;
    }
  }
  return null;
};
