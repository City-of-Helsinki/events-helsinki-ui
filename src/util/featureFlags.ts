export type FeatureFlags = {
  SHOW_SIMILAR_EVENTS: boolean;
  EVENTS_HELSINKI_2: boolean;
};

export const getFeatureFlags = (): FeatureFlags => ({
  SHOW_SIMILAR_EVENTS: process.env.REACT_APP_SHOW_SIMILAR_EVENTS === 'true',
  EVENTS_HELSINKI_2: process.env.REACT_APP_EVENTS_HELSINKI_2 === 'true',
});

export const isFeatureEnabled = (feature: keyof FeatureFlags): boolean =>
  getFeatureFlags()[feature];
