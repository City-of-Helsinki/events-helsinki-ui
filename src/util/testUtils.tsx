import { fireEvent } from '@testing-library/react';

export const arrowUpKeyPressHelper = () =>
  fireEvent.keyDown(document, { code: 38, key: 'ArrowUp' });

export const arrowDownKeyPressHelper = () =>
  fireEvent.keyDown(document, { code: 40, key: 'ArrowDown' });

export const escKeyPressHelper = () =>
  fireEvent.keyDown(document, { code: 27, key: 'Escape' });

export const tabKeyPressHelper = () =>
  fireEvent.keyDown(document, { code: 9, key: 'Tab' });
