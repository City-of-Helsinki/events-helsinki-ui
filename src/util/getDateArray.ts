import { DateArray } from 'ics';

import { formatDate } from './dateUtils';

/**
 * Return date as a array of numbers. Used to convert date to a format that createEvent (ics) supports
 * @param date {string}
 * @return {number[]}
 */
export default (date: string): DateArray => {
  const dateArray = formatDate(new Date(date), 'yyyy-M-d-H-m')
    .split('-')
    .map((e) => Number(e));
  return [dateArray[0], dateArray[1], dateArray[2], dateArray[3], dateArray[4]];
};
