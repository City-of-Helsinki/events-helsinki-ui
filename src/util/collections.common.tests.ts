import { MockedResponse } from '@apollo/react-testing';
import { screen, waitFor } from '@testing-library/react';
import range from 'lodash/range';

import {
  CollectionListDocument,
  CollectionListQueryVariables,
  CollectionListResponse,
} from '../generated/graphql';
import { fakeCollections } from './mockDataUtils';
import { render } from './testUtils';

export const getCollectionQueryListMocks = (
  collections: CollectionListResponse,
  variables?: CollectionListQueryVariables
): MockedResponse[] => [
  {
    request: {
      query: CollectionListDocument,
      variables,
    },
    result: { data: { collectionList: collections } },
  },
];

const renderAndWaitForComponentToBeLoaded = async ({
  component,
  collections,
  mocks = [],
  variables,
}: {
  component: React.ReactElement;
  collections: CollectionListResponse;
  mocks: MockedResponse[];
  variables?: CollectionListQueryVariables;
}) => {
  const collectionMocks = getCollectionQueryListMocks(collections, variables);
  render(component, { mocks: [...collectionMocks, ...mocks] });

  await waitFor(() => {
    expect(screen.queryByTestId('loading-spinner')).not.toBeInTheDocument();
  });
};

const renderComponentAndExpectOddCollectionsToBePresent = async ({
  component,
  collections,
  collectionNames,
  mocks = [],
  variables,
}: {
  component: React.ReactElement;
  collections: CollectionListResponse;
  collectionNames: string[];
  mocks?: MockedResponse[];
  variables?: CollectionListQueryVariables;
}) => {
  await renderAndWaitForComponentToBeLoaded({
    component,
    collections,
    mocks,
    variables,
  });

  collections.data.forEach((collection, index) => {
    if (index % 2) {
      expect(
        screen.queryByText(collectionNames[index])
      ).not.toBeInTheDocument();
    } else {
      expect(screen.getByText(collectionNames[index])).toBeInTheDocument();
    }
  });
};

export const collectionListFilterTests = ({
  component,
  generatedCollectionListSize = 7,
  mocks = [],
  variables,
}: {
  component: React.ReactElement;
  generatedCollectionListSize?: number;
  mocks?: MockedResponse[];
  variables?: CollectionListQueryVariables;
}): void => {
  const collectionIds = range(generatedCollectionListSize).map((id) =>
    (id + 1).toString()
  );
  const collectionNames = collectionIds.map((id) => `Collection ${id}`);

  it('should show all collections', async () => {
    const collections = fakeCollections(
      collectionIds.length,
      collectionIds.map((id, index) => ({
        id,
        title: { fi: collectionNames[index] },
      }))
    );
    await renderAndWaitForComponentToBeLoaded({
      component,
      collections,
      mocks,
      variables,
    });

    collections.data.forEach((collection) => {
      expect(screen.getByText(collection.title.fi)).toBeInTheDocument();
    });
  });

  it('should not show collection if language is not supported', async () => {
    const collections = fakeCollections(
      collectionIds.length,
      collectionIds.map((id, index) => ({
        id,
        title: { fi: Boolean(1 - (index % 2)) ? collectionNames[index] : '' },
      }))
    );

    await renderComponentAndExpectOddCollectionsToBePresent({
      component,
      collections,
      collectionNames,
      mocks,
      variables,
    });
  });

  it('should not show expired collections', async () => {
    const collections = fakeCollections(
      collectionIds.length,
      collectionIds.map((id, index) => ({
        id,
        expired: Boolean(index % 2),
        title: { fi: collectionNames[index] },
      }))
    );

    await renderComponentAndExpectOddCollectionsToBePresent({
      component,
      collections,
      collectionNames,
      mocks,
      variables,
    });
  });

  it('should not show collections that are not live', async () => {
    const collections = fakeCollections(
      collectionIds.length,
      collectionIds.map((id, index) => ({
        id,
        live: Boolean(1 - (index % 2)),
        title: { fi: collectionNames[index] },
      }))
    );
    await renderComponentAndExpectOddCollectionsToBePresent({
      component,
      collections,
      collectionNames,
      mocks,
      variables,
    });
  });
};
