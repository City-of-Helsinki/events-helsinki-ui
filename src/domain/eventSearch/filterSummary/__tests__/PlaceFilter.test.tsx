import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import * as React from 'react';

import { PlaceDetailsDocument } from '../../../../generated/graphql';
import { render } from '../../../../util/testUtils';
import mockPlace from '../../../place/__mocks__/place';
import PlaceFilter from '../PlaceFilter';

const mocks = [
  {
    request: {
      query: PlaceDetailsDocument,
      variables: {
        id: mockPlace.id,
      },
    },
    result: {
      data: {
        placeDetails: mockPlace,
      },
    },
  },
];

test('matches snapshot', async () => {
  const { container } = render(
    <PlaceFilter id={mockPlace.id || ''} onRemove={jest.fn()} />,
    { mocks }
  );

  await screen.findByText((mockPlace.name || {})['fi'] || '');
  expect(container.firstChild).toMatchSnapshot();
});

it('calls onRemove callback when remove button is clicked ', async () => {
  const onClickMock = jest.fn();
  render(<PlaceFilter id={mockPlace.id || ''} onRemove={onClickMock} />, {
    mocks,
  });

  await screen.findByText((mockPlace.name || {})['fi'] || '');

  userEvent.click(screen.getByRole('button'));

  expect(onClickMock).toHaveBeenCalled();
  expect(onClickMock).toHaveBeenCalledWith(mockPlace.id, 'place');
});
