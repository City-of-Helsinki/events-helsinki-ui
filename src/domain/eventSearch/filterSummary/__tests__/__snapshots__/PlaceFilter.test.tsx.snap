// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  class="filter place"
>
  <button
    aria-label="Poista suodatin: Gräsan taitojen talo"
    class="closeButton"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="Icon-module_icon__1Jtzj icon_hds-icon__1YqNC Icon-module_s__2WGWe icon_hds-icon--size-s__2Lkik"
      role="img"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <g
        fill="none"
        fill-rule="evenodd"
      >
        <rect
          height="24"
          width="24"
        />
        <polygon
          fill="currentColor"
          points="18 7.5 13.5 12 18 16.5 16.5 18 12 13.5 7.5 18 6 16.5 10.5 12 6 7.5 7.5 6 12 10.5 16.5 6"
        />
      </g>
    </svg>
  </button>
  Gräsan taitojen talo
</div>
`;
